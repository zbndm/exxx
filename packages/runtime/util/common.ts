export function isKeyof<T extends Record<string, unknown>>(
  obj: T,
  possibleKey: string | number | symbol
): possibleKey is keyof T {
  return possibleKey in obj;
}

export function lookupUpperCase<T>(dict: Record<string, T>, key: string): T;
export function lookupUpperCase<T>(dict: Record<string, T>, key: string): null;
export function lookupUpperCase<T>(dict: Record<string, T>, key: string) {
  key = key.toUpperCase();
  if (isKeyof(dict, key)) {
    return dict[key];
  }

  return null;
}
