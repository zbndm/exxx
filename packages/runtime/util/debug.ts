export const intro =
  process.env.NODE_ENV === 'development'
    ? () => {
        console.log(
          '%c%s',
          'font-size: 21px; color: cyan; text-shadow: 1px 1px 0 black, 1px -1px 0 black, -1px 1px 0 black, -1px -1px 0 black',
          'Running in development mode.\nClick below to open the popup page:'
        );
        console.log(
          '%c%s',
          'font-size: 15px',
          browser.runtime.getURL('popup.html')
        );
      }
    : () => {
        // ignore in test & production
      };

export const log: Console['log'] = console.log.bind(globalThis);

export const error: Console['error'] = console.error.bind(globalThis);

export const assert: Console['assert'] = console.assert.bind(globalThis);

type AssertNever = (message: string) => never;
export const never = console.assert.bind(globalThis, true) as AssertNever;
