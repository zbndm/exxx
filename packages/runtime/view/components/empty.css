.empty {
  max-width: 360px;
}

.empty__header {
  padding: 16px 24px;
  background-color: #00bbff;
  color: rgba(255, 255, 255, 0.87);
}

.empty__header ::selection {
  background-color: #da2d76;
}

.empty__logo {
  height: 8px;
  opacity: 0.8;
  padding: 8px 0;
}

.empty__info {
  padding: 16px 0 0 0;
}

.empty__intro {
  padding: 16px 24px;
}
