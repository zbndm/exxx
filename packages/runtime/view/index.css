body {
  padding: 0;
  margin: 0;
  font-family: 'Segoe UI', Segoe, Tahoma, Geneva, sans-serif;
  font-size: 16px;
  min-width: 360px;
  max-width: 360px;
  background-color: #f2f5f7;
  color: #829ca9;
}

#react-root {
  min-height: 100px;
}
