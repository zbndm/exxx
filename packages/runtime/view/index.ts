import * as emojiService from '../services/emoji/emoji_service';
import * as preferenceService from '../services/preference/preference_service';

import './index.css';

export const start = (callback?: () => void) =>
  Promise.all([preferenceService.init(), emojiService.init()]).then(callback);
