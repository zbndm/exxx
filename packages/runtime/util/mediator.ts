import * as debug from './debug';

interface Command {
  execute(...args: unknown[]): Promise<unknown> | void;
}

type CommandResult<C extends Command> = Unwrap<ReturnType<C['execute']>>;

export interface Action<C extends Command> {
  (command: C, executionResult: CommandResult<C>): Promise<void> | void;
}

interface Observer<C extends Command> {
  commandCtor: new (...args: any[]) => C;
  action: Action<C>;
}

export const observers: Observer<Command>[] = [];

export function subscribe<C extends Command>(
  commandCtor: new (...args: any[]) => C,
  action: Action<C>
) {
  debug.assert(typeof action === 'function');

  observers.push({
    commandCtor,
    action: action as Action<Command>,
  });
}

export function publish(command: Command) {
  async function _publish(promise: Promise<unknown> | void) {
    let result: unknown;
    try {
      result = await promise;
    } catch (err) {
      debug.error(err);
      return;
    }

    return Promise.all(
      observers.map(async function (observer) {
        try {
          if (command instanceof observer.commandCtor) {
            return await observer.action(command, result);
          }
        } catch (err) {
          debug.error(err);
        }
      })
    );
  }

  return _publish(command.execute());
}
