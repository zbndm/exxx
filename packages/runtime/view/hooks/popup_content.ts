import { useEffect, useState } from 'react';

import { Report } from '../../lib/report';
import * as xpc from '../../services/xpc/xpc_popup_service';
import * as debug from '../../util/debug';

import { useQueryParams } from './query_params';

interface PopupContentInit {
  state: null;
}
interface PopupContentSuccess {
  state: 'success';
  report: Report;
}
interface PopupContentFailure {
  state: 'failure';
  error: Error;
}

type PopupContent =
  | PopupContentInit
  | PopupContentSuccess
  | PopupContentFailure;

export function usePopupContent() {
  const [response, setResponse] = useState<PopupContent>({
    state: null,
  });

  const TAG = location.search;

  const { tab } = useQueryParams();
  const tabId = typeof tab === 'string' ? parseInt(tab) : null;

  useEffect(() => {
    let shouldIgnore = false;

    debug.log(`Popup${TAG}: Fetching report`);

    if (!tabId) {
      return;
    }

    xpc.dispatch('fetchReport', tabId).then(
      (report) => {
        debug.log(`Popup${TAG}: Report has been fetched`, report);

        if (shouldIgnore) {
          return;
        }

        if (report) {
          setResponse({
            report,
            state: 'success',
          });
        } // FIXME else throw a "Not Found"
      },
      (error: Error) => {
        debug.error(error);

        if (shouldIgnore) {
          return;
        }

        setResponse({
          state: 'failure',
          error,
        });
      }
    );

    return () => {
      shouldIgnore = true;
    };
  }, [tabId]);

  return response;
}
