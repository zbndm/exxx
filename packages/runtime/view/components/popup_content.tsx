import { hot } from 'react-hot-loader/root';
// eslint-disable-next-line import/order
import { FC } from 'react';

import { usePopupContent } from '../hooks/popup_content';

import { Empty } from './empty';
import { HeaderPrimary } from './header_primary';
import { Traceroute } from './traceroute';

export const _PopupContent: FC = () => {
  const popupResponse = usePopupContent();

  if (popupResponse.state === 'success' && popupResponse.report) {
    return (
      <>
        <HeaderPrimary report={popupResponse.report} />
        <Traceroute {...popupResponse.report} />
      </>
    );
  } else {
    return <Empty />;
  }
};

export const PopupContent = hot(_PopupContent);
