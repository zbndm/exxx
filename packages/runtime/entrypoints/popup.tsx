import { render } from 'react-dom';

import * as debug from '../util/debug';
import { start } from '../view';
import { PopupContent } from '../view/components/popup_content';

debug.intro();

void start(() =>
  render(<PopupContent />, document.getElementById('react-root'))
);
