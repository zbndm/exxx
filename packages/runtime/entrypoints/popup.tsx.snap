// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup when rendering for a CloudFront-enabled website should render country info for CloudFront Edge IP 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        143.204.202.99
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=143.204.202.99"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=143.204.202.99"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "NA",
        "extra": Object {
          "registeredCountry": "US",
        },
        "heuristic": "ip",
        "isoCountry": "US",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="US"
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
          style={
            Object {
              "background": "url('/assets/twemoji/us.svg') no-repeat",
            }
          }
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
            data-flag="ðŸ‡ºðŸ‡¸"
            style={
              Object {
                "background": "url('/assets/twemoji/us.svg') no-repeat",
              }
            }
          >
            Country:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
      <CountryItem
        countryCode="US"
        shouldRenderFlag={false}
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          size="xs"
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
          >
            Legal registration:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "Server",
            "value": "Apache",
          },
          Object {
            "name": "X-Cache",
            "value": "Miss from cloudfront",
          },
          Object {
            "name": "Via",
            "value": "1.1 1764af62d635a1a6ee51aabc37405452.cloudfront.net (CloudFront)",
          },
          Object {
            "name": "X-Amz-Cf-Pop",
            "value": "FRA53-C1",
          },
          Object {
            "name": "X-Amz-Cf-Id",
            "value": "XXWLOSXVpyCPH_qWH_e61SK44oFpgwRyAst9DnHWCKlxjC8Dm6f7uw==",
          },
        ],
        "ip": "143.204.202.99",
        "url": "https://www.workday.com/",
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Server Software: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              Apache
            </span>
          </Component>
        </p>
      </Component>
    </div>
  </ServerInfo>
</div>
`;

exports[`Popup when rendering for a CloudFront-enabled website should show CloudFront POP abbreviation 1`] = `
<div
  className="cloudfront-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      AWS CloudFront routed through
    </p>
  </Component>
  <div
    className="traceroute__nested"
  >
    <CountryItem
      countryCode="DE"
    >
      <Component
        className="country-item country-item_twemoji country-item_with-flag"
        data-flag="ðŸ‡©ðŸ‡ª"
        size="xs"
        style={
          Object {
            "background": "url('/assets/twemoji/de.svg') no-repeat",
          }
        }
      >
        <p
          className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡©ðŸ‡ª"
          style={
            Object {
              "background": "url('/assets/twemoji/de.svg') no-repeat",
            }
          }
        >
          <strong>
            FRA
          </strong>
           data center in
          <br />
          <strong>
            Germany
          </strong>
        </p>
      </Component>
    </CountryItem>
    <Component
      size="xs"
      variant="light"
    >
      <p
        className="typography typography_light typography_xs"
      >
        Cache: 
        <Component
          variant="mono"
        >
          <span
            className="typography typography_mono"
          >
            Miss from cloudfront
          </span>
        </Component>
      </p>
    </Component>
  </div>
</div>
`;

exports[`Popup when rendering for a Cloudflare-protected website should render country info for Cloudflare IP 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        104.27.165.61
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.27.165.61"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.27.165.61"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "NA",
        "extra": null,
        "heuristic": "ip",
        "isoCountry": "US",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="US"
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
          style={
            Object {
              "background": "url('/assets/twemoji/us.svg') no-repeat",
            }
          }
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
            data-flag="ðŸ‡ºðŸ‡¸"
            style={
              Object {
                "background": "url('/assets/twemoji/us.svg') no-repeat",
              }
            }
          >
            Country:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "cf-ray",
            "value": "5d5d5d5d5d5d5d5d-CPH",
          },
          Object {
            "name": "cf-cache-status",
            "value": "REVALIDATED",
          },
          Object {
            "name": "cf-request-id",
            "value": "00000004f10000000000000000000001",
          },
          Object {
            "name": "x-nf-request-id",
            "value": "b1b1b1b1-c9c9-4545-afaf-292929292929-33333333",
          },
          Object {
            "name": "server",
            "value": "cloudflare",
          },
        ],
        "ip": "104.27.165.61",
        "url": "https://ilin.dk/",
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Server Software: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              cloudflare
            </span>
          </Component>
        </p>
      </Component>
    </div>
  </ServerInfo>
</div>
`;

exports[`Popup when rendering for a Cloudflare-protected website should show Cloudflare data center abbreviation 1`] = `
<div
  className="cloudflare-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      Cloudflare anycast rerouted through
    </p>
  </Component>
  <div
    className="traceroute__nested"
  >
    <CountryItem
      countryCode="DK"
    >
      <Component
        className="country-item country-item_twemoji country-item_with-flag"
        data-flag="ðŸ‡©ðŸ‡°"
        size="xs"
        style={
          Object {
            "background": "url('/assets/twemoji/dk.svg') no-repeat",
          }
        }
      >
        <p
          className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡©ðŸ‡°"
          style={
            Object {
              "background": "url('/assets/twemoji/dk.svg') no-repeat",
            }
          }
        >
          <strong>
            CPH
          </strong>
           data center in
          <br />
          <strong>
            Denmark
          </strong>
        </p>
      </Component>
    </CountryItem>
    <Component
      size="xs"
      variant="light"
    >
      <p
        className="typography typography_light typography_xs"
      >
        Cache Status was 
        <Component
          variant="mono"
        >
          <span
            className="typography typography_mono"
          >
            REVALIDATED
          </span>
        </Component>
         (
        <Component
          href="https://support.cloudflare.com/hc/articles/200172516-Understanding-Cloudflare-s-CDN#h_bd959d6a-39c0-4786-9bcd-6e6504dcdb97"
          size="xs"
          target="_blank"
          title="What does \\"Cache status\\" mean?"
          variant="light"
        >
          <a
            className="typography typography_light typography_xs link"
            href="https://support.cloudflare.com/hc/articles/200172516-Understanding-Cloudflare-s-CDN#h_bd959d6a-39c0-4786-9bcd-6e6504dcdb97"
            target="_blank"
            title="What does \\"Cache status\\" mean?"
          >
            ?
          </a>
        </Component>
        )
      </p>
    </Component>
  </div>
</div>
`;

exports[`Popup when rendering for a Fastly-accelerated website should render country info for Fastly node IP 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        151.101.1.195
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=151.101.1.195"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=151.101.1.195"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "NA",
        "extra": Object {
          "registeredCountry": "US",
        },
        "heuristic": "ip",
        "isoCountry": "US",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="US"
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
          style={
            Object {
              "background": "url('/assets/twemoji/us.svg') no-repeat",
            }
          }
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
            data-flag="ðŸ‡ºðŸ‡¸"
            style={
              Object {
                "background": "url('/assets/twemoji/us.svg') no-repeat",
              }
            }
          >
            Country:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
      <CountryItem
        countryCode="US"
        shouldRenderFlag={false}
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          size="xs"
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
          >
            Legal registration:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "x-served-by",
            "value": "cache-cph20643-CPH",
          },
          Object {
            "name": "x-cache",
            "value": "MISS",
          },
          Object {
            "name": "x-cache-hits",
            "value": "0",
          },
          Object {
            "name": "x-fastly-request-id",
            "value": "d252eca9b9d3a9449ecff9e552be6dd9f561a360",
          },
        ],
        "ip": "151.101.1.195",
        "url": "https://example.com/",
      }
    }
  />
</div>
`;

exports[`Popup when rendering for a Fastly-accelerated website should show Fastly POP abbreviation 1`] = `
<div
  className="fastly-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      Fastly rerouted through
    </p>
  </Component>
  <div
    className="traceroute__nested"
  >
    <CountryItem
      countryCode="DK"
    >
      <Component
        className="country-item country-item_twemoji country-item_with-flag"
        data-flag="ðŸ‡©ðŸ‡°"
        size="xs"
        style={
          Object {
            "background": "url('/assets/twemoji/dk.svg') no-repeat",
          }
        }
      >
        <p
          className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡©ðŸ‡°"
          style={
            Object {
              "background": "url('/assets/twemoji/dk.svg') no-repeat",
            }
          }
        >
          <strong>
            CPH
          </strong>
           data center in
          <br />
          <strong>
            Denmark
          </strong>
        </p>
      </Component>
    </CountryItem>
    <Component
      size="xs"
      variant="light"
    >
      <p
        className="typography typography_light typography_xs"
      >
        Cache Status was 
        <Component
          variant="mono"
        >
          <span
            className="typography typography_mono"
          >
            MISS
          </span>
        </Component>
         (
        <Component
          href="https://developer.fastly.com/learning/concepts/shielding/#debugging"
          size="xs"
          target="_blank"
          title="What does \\"Cache status\\" mean?"
          variant="light"
        >
          <a
            className="typography typography_light typography_xs link"
            href="https://developer.fastly.com/learning/concepts/shielding/#debugging"
            target="_blank"
            title="What does \\"Cache status\\" mean?"
          >
            ?
          </a>
        </Component>
        )
      </p>
    </Component>
  </div>
</div>
`;

exports[`Popup when rendering for a Netherlands-based website with legal entity in the US should render IP country info 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        40.114.177.156
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=40.114.177.156"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=40.114.177.156"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "EU",
        "extra": Object {
          "registeredCountry": "US",
        },
        "heuristic": "ip",
        "isoCountry": "NL",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="NL"
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡³ðŸ‡±"
          size="xs"
          style={
            Object {
              "background": "url('/assets/twemoji/nl.svg') no-repeat",
            }
          }
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
            data-flag="ðŸ‡³ðŸ‡±"
            style={
              Object {
                "background": "url('/assets/twemoji/nl.svg') no-repeat",
              }
            }
          >
            Country:
            <br />
            <strong>
              Netherlands
            </strong>
          </p>
        </Component>
      </CountryItem>
      <CountryItem
        countryCode="US"
        shouldRenderFlag={true}
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
          style={
            Object {
              "background": "url('/assets/twemoji/us.svg') no-repeat",
            }
          }
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
            data-flag="ðŸ‡ºðŸ‡¸"
            style={
              Object {
                "background": "url('/assets/twemoji/us.svg') no-repeat",
              }
            }
          >
            Legal registration:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "status",
            "value": "304",
          },
          Object {
            "name": "server",
            "value": "nginx",
          },
        ],
        "ip": "40.114.177.156",
        "url": "https://duckduckgo.com/",
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Server Software: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              nginx
            </span>
          </Component>
        </p>
      </Component>
    </div>
  </ServerInfo>
</div>
`;

exports[`Popup when rendering for a Switzerland-based website should render IP country info 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        185.70.41.35
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=185.70.41.35"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=185.70.41.35"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "EU",
        "extra": Object {
          "registeredCountry": "CH",
        },
        "heuristic": "ip",
        "isoCountry": "CH",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="CH"
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡¨ðŸ‡­"
          size="xs"
          style={
            Object {
              "background": "url('/assets/twemoji/ch.svg') no-repeat",
            }
          }
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
            data-flag="ðŸ‡¨ðŸ‡­"
            style={
              Object {
                "background": "url('/assets/twemoji/ch.svg') no-repeat",
              }
            }
          >
            Country:
            <br />
            <strong>
              Switzerland
            </strong>
          </p>
        </Component>
      </CountryItem>
      <CountryItem
        countryCode="CH"
        shouldRenderFlag={false}
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          size="xs"
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
          >
            Legal registration:
            <br />
            <strong>
              Switzerland
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": undefined,
        "ip": "185.70.41.35",
        "url": "https://protonmail.com/",
      }
    }
  />
</div>
`;

exports[`Popup when rendering for a website behind Cloudflare with Powered By info exposed should render country info for Cloudflare IP 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        104.26.7.76
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.26.7.76"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.26.7.76"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "NA",
        "extra": null,
        "heuristic": "ip",
        "isoCountry": "US",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="US"
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
          style={
            Object {
              "background": "url('/assets/twemoji/us.svg') no-repeat",
            }
          }
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
            data-flag="ðŸ‡ºðŸ‡¸"
            style={
              Object {
                "background": "url('/assets/twemoji/us.svg') no-repeat",
              }
            }
          >
            Country:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "x-powered-by",
            "value": "WP Engine",
          },
          Object {
            "name": "x-cacheable",
            "value": "SHORT",
          },
          Object {
            "name": "x-cache",
            "value": "HIT: 8",
          },
          Object {
            "name": "x-cache-group",
            "value": "normal",
          },
          Object {
            "name": "x-orig-cache-control",
            "value": "max-age=0",
          },
          Object {
            "name": "cf-cache-status",
            "value": "DYNAMIC",
          },
          Object {
            "name": "cf-request-id",
            "value": "050505050500000ebebebebeb0000001",
          },
          Object {
            "name": "server",
            "value": "cloudflare",
          },
          Object {
            "name": "cf-ray",
            "value": "5e5e5e5e5e5e5e5e7-FRA",
          },
        ],
        "ip": "104.26.7.76",
        "url": "https://peakon.com/",
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Server Software: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              cloudflare
            </span>
          </Component>
        </p>
      </Component>
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Powered By: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              WP Engine
            </span>
          </Component>
        </p>
      </Component>
    </div>
  </ServerInfo>
</div>
`;

exports[`Popup when rendering for a website behind Cloudflare with Powered By info exposed should render the header value to the user 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        104.26.7.76
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.26.7.76"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.26.7.76"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "NA",
        "extra": null,
        "heuristic": "ip",
        "isoCountry": "US",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="US"
      >
        <Component
          className="country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
          style={
            Object {
              "background": "url('/assets/twemoji/us.svg') no-repeat",
            }
          }
        >
          <p
            className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
            data-flag="ðŸ‡ºðŸ‡¸"
            style={
              Object {
                "background": "url('/assets/twemoji/us.svg') no-repeat",
              }
            }
          >
            Country:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "x-powered-by",
            "value": "WP Engine",
          },
          Object {
            "name": "x-cacheable",
            "value": "SHORT",
          },
          Object {
            "name": "x-cache",
            "value": "HIT: 8",
          },
          Object {
            "name": "x-cache-group",
            "value": "normal",
          },
          Object {
            "name": "x-orig-cache-control",
            "value": "max-age=0",
          },
          Object {
            "name": "cf-cache-status",
            "value": "DYNAMIC",
          },
          Object {
            "name": "cf-request-id",
            "value": "050505050500000ebebebebeb0000001",
          },
          Object {
            "name": "server",
            "value": "cloudflare",
          },
          Object {
            "name": "cf-ray",
            "value": "5e5e5e5e5e5e5e5e7-FRA",
          },
        ],
        "ip": "104.26.7.76",
        "url": "https://peakon.com/",
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Server Software: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              cloudflare
            </span>
          </Component>
        </p>
      </Component>
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Powered By: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              WP Engine
            </span>
          </Component>
        </p>
      </Component>
    </div>
  </ServerInfo>
</div>
`;

exports[`Popup when rendering for a website behind Cloudflare with Powered By info exposed should show Cloudflare data center abbreviation 1`] = `
<div
  className="cloudflare-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      Cloudflare anycast rerouted through
    </p>
  </Component>
  <div
    className="traceroute__nested"
  >
    <CountryItem
      countryCode="DE"
    >
      <Component
        className="country-item country-item_twemoji country-item_with-flag"
        data-flag="ðŸ‡©ðŸ‡ª"
        size="xs"
        style={
          Object {
            "background": "url('/assets/twemoji/de.svg') no-repeat",
          }
        }
      >
        <p
          className="typography typography_xs country-item country-item_twemoji country-item_with-flag"
          data-flag="ðŸ‡©ðŸ‡ª"
          style={
            Object {
              "background": "url('/assets/twemoji/de.svg') no-repeat",
            }
          }
        >
          <strong>
            FRA
          </strong>
           data center in
          <br />
          <strong>
            Germany
          </strong>
        </p>
      </Component>
    </CountryItem>
    <Component
      size="xs"
      variant="light"
    >
      <p
        className="typography typography_light typography_xs"
      >
        Cache Status was 
        <Component
          variant="mono"
        >
          <span
            className="typography typography_mono"
          >
            DYNAMIC
          </span>
        </Component>
         (
        <Component
          href="https://support.cloudflare.com/hc/articles/200172516-Understanding-Cloudflare-s-CDN#h_bd959d6a-39c0-4786-9bcd-6e6504dcdb97"
          size="xs"
          target="_blank"
          title="What does \\"Cache status\\" mean?"
          variant="light"
        >
          <a
            className="typography typography_light typography_xs link"
            href="https://support.cloudflare.com/hc/articles/200172516-Understanding-Cloudflare-s-CDN#h_bd959d6a-39c0-4786-9bcd-6e6504dcdb97"
            target="_blank"
            title="What does \\"Cache status\\" mean?"
          >
            ?
          </a>
        </Component>
        )
      </p>
    </Component>
  </div>
</div>
`;
